editor/editor_help.cpp

```diff
@@ -2935,7 +2948,64 @@ void EditorHelp::_gen_extensions_docs() {
 	}
 }
 
-void EditorHelp::generate_doc(bool p_use_cache) {
+void EditorHelp::_load_script_doc_cache() {
+	if (ResourceLoader::exists(get_script_doc_cache_full_path())) {
+		Ref<Resource> script_doc_cache_res = ResourceLoader::load(get_script_doc_cache_full_path());
+		if (script_doc_cache_res.is_valid()) {
+			Array classes = script_doc_cache_res->get_meta("classes", Array());
+			for (int i = 0; i < classes.size(); i++)
+				doc->add_doc(DocData::ClassDoc::from_dict(classes[i]));
+		}
+		return;
+	}
+
+	// Force compile all scripts, which should generate their documentation.
+	WARN_PRINT_ED("Script documentation cache not found. Regenerating it may take a while.");
+	HashMap<String, bool> is_extension_script; // Caches ScriptServer results so it's not being queried for every file, only for every extension.
+	List<Ref<DirAccess>> dirs;
+	dirs.push_front(DirAccess::create(DirAccess::ACCESS_RESOURCES));
+	while (!dirs.is_empty()) {
+		Ref<DirAccess> dir = dirs.front()->get();
+		dirs.pop_front();
+		for (const String file : dir->get_files()) {
+			if (file.begins_with("."))
+				continue;
+			if (!is_extension_script.has(file.get_extension()))
+				is_extension_script[file.get_extension()] = ScriptServer::get_language_for_extension(file.get_extension()) != nullptr;
+
+			if (is_extension_script[file.get_extension()]) {
+				Ref<Script> scr = ResourceLoader::load(file);
+				if (scr.is_valid())
+					for (DocData::ClassDoc cd : scr->get_documentation())
+						doc->add_doc(cd);
+			}
+		}
+
+		for (const String subdir : dir->get_directories()) {
+			if (!subdir.begins_with("."))
+				dirs.push_front(DirAccess::open(DirAccess::get_full_path(dir->get_current_dir() + "/" + subdir, DirAccess::ACCESS_RESOURCES)));
+		}
+	}
+
+	save_script_doc_cache();
+}
```

Comment:
This goes through every single file in `res://` and forces it to recompile so we can get its documentation. I think this is the most efficient I can make it if the cache is not present. But I'd love more eyes on this to see if I'm missing something.

Alternatively, this can be done somewhere in `EditorFileSystem`.

Reply:
For posterity, I did try moving this code into `EditorFileSystem`, but we run into threading issues. If we call `EditorFileSystem` and use existing methods like `_update_script_documentation()`, then we cannot be running from `EditorHelp::worker_thread`, which we are and want to.

I don't think it's worth the threading hassle, we can just have walk through the filesystem instead of using the project's file cache :)

Reply:
This has been changed to waiting until `EditorFileSystem` is done with its scan and using its datastructures.